src/adapter/src/optimize/dataflow_expiration.rs

```diff
@@ -0,0 +1,146 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+
+//! Helper function for dataflow expiration checks.
+//!
+//! A [`TimeDependence`] describes how a dataflow follows wall-clock time, and this module offers
+//! functions to compute the time dependence based on the current state of the catalog.
+//!
+//! For a set of global IDs, the helper determines how a dataflow that depends on the set of IDs
+//! follows wall-clock time. Within the catalog, we remember how all installed objects follow
+//! wall-clock time, which turns this step into gathering and combining different time dependencies:
+//! * A meet of anything with wall-clock time results in wall-clock time.
+//! * A meet of anything but wall-clock time and a refresh schedule results in a refresh schedule
+//!   that depends on the deduplicated collection of dependencies.
+//! * Otherwise, a dataflow is indeterminate, which expresses that we either don't know how it
+//!   follows wall-clock time, or is a constant collection.
+//!
+//! The time dependence needs to be computed on the actual dependencies, and not on catalog
+//! uses. An optimized dataflow depends on concrete indexes, and has unnecessary dependencies
+//! pruned. Additionally, transitive dependencies can depend on indexes that do not exist anymore,
+//! which makes combining run-time information with catalog-based information inconclusive.
+
+use mz_catalog::memory::objects::{CatalogItem, DataSourceDesc, TableDataSource};
+use mz_compute_types::time_dependence::TimeDependence;
+use mz_ore::soft_panic_or_log;
+use mz_repr::refresh_schedule::RefreshSchedule;
+use mz_repr::GlobalId;
+use mz_sql::catalog::CatalogItem as _;
+use mz_storage_types::sources::{GenericSourceConnection, Timeline};
+
+use crate::catalog::Catalog;
+
+/// Determine a time dependence based on a set of global IDs. The result includes the schedule if
+/// supplied and needed. Returns `None` if called with an empty iterator, or if all dependencies
+/// are indeterminate.
+pub(crate) fn time_dependence(
+    catalog: &Catalog,
+    ids: impl IntoIterator<Item = GlobalId>,
+    schedule: Option<RefreshSchedule>,
+) -> Option<TimeDependence> {
+    // Collect all time dependencies of our dependencies.
+    let mut time_dependencies = ids
+        .into_iter()
+        .filter_map(|id| time_dependence_for_id(catalog, id))
+        .collect::<Vec<_>>();
+
+    // Sort and dedupe to remove redundancy.
+    time_dependencies.sort();
+    time_dependencies.dedup();
+
+    let time_dependence = if time_dependencies
+        .iter()
+        .any(|dep| *dep == TimeDependence::default())
+    {
+        // Wall-clock dependency is dominant.
+        Some(TimeDependence::new(schedule, vec![]))
```

Comment:
I'm wondering if this is also true for CTs. Seems like with the current implementation you could create a CT that has as its input a `REFRESH EVERY (1 day)` MV and a plain old table as reference. The table moves in wallclock time, but the CT's output only changes when the input changes, so if the input frontier jumps forward 1day, the CT's frontier can too? Not sure if that's the case with the current implementation though.

Also, if we decide to go with the ["commit at every timestamp" option](https://www.notion.so/materialize/Continual-Tasks-Decision-Log-7279978e873741ca83d0c2f8382dad5f?pvs=4#6a3268e42b604439b05d097ff4e3f076), the above reasoning doesn't apply.

Reply:
Removed for continual tasks.